### PR TITLE
 Update DLAMI BASE AMI Logic to switch between OSS and Proprietary Nvidia Driver AMI

### PR DESCRIPTION
**Pull request by sirutBuasai**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Tests run

- PyTorch 2.2 Training EC2 image tests
  - [e32684a](https://github.com/aws/deep-learning-containers/pull/3760/commits/e32684a77aaf3a07c75a1d36bb4eafedee9967a8)
    - ECS test passed
    - EKS test passed
    - Sanity test passed
    - EC2 test passed
- PyTorch 2.2 Training SM image tests
  - [ac78c1f](https://github.com/aws/deep-learning-containers/pull/3760/commits/ac78c1fe183454d137b18f9a5b22e9d8b7913e7a)
    - ECS test passed
    - EKS test passed
    - Sanity test passed
    - EC2 test passed
    - SM benchmark test passed
    - SM efa test passed
    - SM local test passed
    - SM remote test passed
    - SM rc test passed
- Tensorflow 2.14 Training EC2/SM image tests
  - [4013f89](https://github.com/aws/deep-learning-containers/pull/3760/commits/4013f890844a676c553b2ec61f8ca5d49f3b4e1f)
    - ECS test passed
    - EKS test passed
    - Sanity test passed
    - EC2 test passed
    - SM benchmark test passed
    - SM efa test passed
    - SM local test passed
    - SM remote test passed
    - SM rc test passed
- PyTorch 2.2 Inference EC2/SM image tests
  - [64d9afa](https://github.com/aws/deep-learning-containers/pull/3760/commits/64d9afae35fb67fa4622ec3eceb767aaef216030)
    - ECS test passed
    - EKS test passed
    - Sanity test passed
    - EC2 test passed
    - SM benchmark test passed
    - SM efa test passed
    - SM local test passed
    - SM remote test passed
    - SM rc test passed

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Click the checkbox to enable a build to execute upon merge.

*Note: By default, pipelines are set to "latest". Replace with major.minor framework version if you do not want "latest".*

- [ ] build_pytorch_training_latest
- [ ] build_pytorch_inference_latest
- [ ] build_tensorflow_training_latest
- [ ] build_tensorflow_inference_latest

</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/__init__.py [hunk 1]
```diff
@@ -2370,3 +2392,95 @@ def get_image_spec_from_buildspec(image_uri, dlc_folder_path):
...
+                                                             "p3dn.24xlarge",
+                                                             "g3s.xlarge",
+                                                             "g3.4xlarge",
+                                                             "g3.8xlarge",
+                                                             "g3.16xlarge",]
+
+    Other instances will default to Proprietary Nvidia Driver DLAMI
+    """
+
+    base_proprietary_dlami_instances = [
```
Comment (arjkesh) [inline test/test_utils/__init__.py:2432]: do we need to hard code the actual instance type, or is it fair to assume a certain instance type based on family (i.e. p3, g3, instead of all of the specific types)

Comment (sirutBuasai) [reply test/test_utils/__init__.py:2432]: We can, I just thought it wouldn't make much of a difference between maintaining a list of family types vs list all all instance names so I opted for the more verbose one in the event of an unlikely edge case such as removing a particular instance type from proprietary dlami.

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training_2_2.py [hunk 2]
```diff
@@ -34,6 +35,7 @@ def test_pytorch_2_2_gpu(
         (common_cases.nvapex, (pytorch_training, ec2_connection)),
         (common_cases.pytorch_training_torchaudio, (pytorch_training, ec2_connection)),
         (common_cases.pytorch_cudnn_match_gpu, (pytorch_training, ec2_connection, region)),
+        (common_cases.pytorch_curand_gpu, (pytorch_training, ec2_connection)),
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training_2_2.py:37]: I checked `testCurand` file and [these lines](https://github.com/aws/deep-learning-containers/blob/3b23162c36308525c207edce1aff5925ce4c5a45/test/dlc_tests/container_tests/bin/testCurand#L9-L11) were changed  in PT 2.0 release to account for the fact that curand is shipped with conda as Lai mentioned.

I don't see a reason for removing this test since it is configured to test both pre and post conda distribution. But if you think it is an unnecessary overhead to maintain this test, then we can remove it.

[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training_2_2.py:37]: You can see the test time at the end of runs, I think it's only a couple seconds. If it can be run on multi-gpu instances without issue then I'm OK keeping it.

File: test/test_utils/__init__.py [hunk 3]
```diff
@@ -2370,3 +2392,95 @@ def get_image_spec_from_buildspec(image_uri, dlc_folder_path):
...
+        "g3.8xlarge",
+        "g3.16xlarge",
+    ]
+
+    # set defaults
+    if linux_dist == "AML2":
+        oss_dlami_us_east_1 = AML2_BASE_OSS_DLAMI_US_EAST_1
+        oss_dlami_us_west_2 = AML2_BASE_OSS_DLAMI_US_WEST_2
+        oss_dlami_name_pattern = (
+            "Deep Learning Base OSS Nvidia Driver AMI (Amazon Linux 2) Version ??.?"
```
Comment (arjkesh) [inline test/test_utils/__init__.py:2448]: can pattern be used for all regions - is there a reason we need to hardcode us-east-1 and us-west-2 throughout?

File: test/test_utils/__init__.py [hunk 4]
```diff
@@ -2370,3 +2392,95 @@ def get_image_spec_from_buildspec(image_uri, dlc_folder_path):
...
+        "p3.16xlarge",
+        "p3dn.24xlarge",
+        "g3s.xlarge",
+        "g3.4xlarge",
+        "g3.8xlarge",
+        "g3.16xlarge",
+    ]
+
+    # set defaults
+    if linux_dist == "AML2":
```
Comment (arjkesh) [inline test/test_utils/__init__.py:2444]: nit: these look more like mappings than variable assignments, I might do something like

```
ami_patterns = {
  "AML2": {
    "oss": <pattern>,
    "proprietary": <pattern>
  },
  "UBUNTU_20": {
    "oss": <pattern>,
    "proprietary": <pattern>
  },
  # i.e. future patterns
  "UBUNTU_22": {
    "oss": <pattern>,
    "proprietary": <pattern>
  },
}
```

File: test/test_utils/__init__.py [hunk 5]
```diff
@@ -2370,3 +2392,95 @@ def get_image_spec_from_buildspec(image_uri, dlc_folder_path):
...
+        )
+
+        proprietary_dlami_us_east_1 = UBUNTU_20_BASE_PROPRIETARY_DLAMI_US_EAST_1
+        proprietary_dlami_us_west_2 = UBUNTU_20_BASE_PROPRIETARY_DLAMI_US_WEST_2
+        proprietary_dlami_name_pattern = (
+            "Deep Learning Base Proprietary Nvidia Driver GPU AMI (Ubuntu 20.04) ????????"
+        )
+
+    instance_ami = (
+        proprietary_dlami_us_east_1
```
Comment (arjkesh) [inline test/test_utils/__init__.py:2470]: nit: this is a very long ternary operator, I think this is going to be difficult to maintain going forward.

If you use the map approach above, you can simplify this expression to

```
ami_type = "proprietary" if instance_type in base_proprietary_dlami_instances else "oss"
return get_ami_id_boto3(region_name=region, ami_name_pattern=ami_patterns[linux_dist][ami_type])
```

File: test/test_utils/__init__.py (hunk 1 above)
Comment (arjkesh) [reply test/test_utils/__init__.py:2432]: How do you plan to keep this list up to date? Is there a reference link you can use (i.e. instead of listening everything out in the function description)?

File: test/test_utils/__init__.py (hunk 3 above)
Comment (sirutBuasai) [reply test/test_utils/__init__.py:2448]: I think this was previously done because `us-east-1` and `us-west-2` are frequently used so we make them a variable so that we don't have to call boto3 api every time we need the AMI and to ensure we use the same AMI through the tests.

Comment (arjkesh) [reply test/test_utils/__init__.py:2448]: if worried about overusing api calls, we can just add a function level cache here. seems essentially what we're trying to mimic with the global assignment approach

File: test/test_utils/__init__.py (hunk 1 above)
Comment (sirutBuasai) [reply test/test_utils/__init__.py:2433]: Added a link for all instance types https://aws.amazon.com/ec2/instance-types/#Accelerated_Computing

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training_2_2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training_2_2.py:37]: Yeah, tests has been passing on multigpu instances. It only lasts 5 second so not much overhead.

File: test/test_utils/__init__.py (hunk 1 above)
Comment (sirutBuasai) [reply test/test_utils/__init__.py:2433]: The current source of truth we have publicly is here https://docs.aws.amazon.com/dlami/latest/devguide/important-changes.html

Jeet has been maintaining this page for most up to date info.

File: test/dlc_tests/conftest.py [hunk 6]
```diff
@@ -564,9 +553,14 @@ def ec2_instance(
         )
         if ec2_instance_ami != PT_GPU_PY3_BENCHMARK_IMAGENET_AMI_US_EAST_1:
             ec2_instance_ami = (
-                AML2_BASE_DLAMI_US_EAST_1
-                if ec2_instance_ami == AML2_BASE_DLAMI_US_WEST_2
-                else UBUNTU_20_BASE_DLAMI_US_EAST_1
+                test_utils.get_instance_type_base_dlami(
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply test/dlc_tests/conftest.py:556]: Yeah this condition was very weird. I don't quite understand it completely but it seemed like the original commit was trying to cast region `us-east-1` though.

To my understanding, it is trying to force `us-east-1` AML2 AMI if the current `ec2_instance_ami` is an AML2 AMI in `us-west-2`. Otherwise, it uses the UL20 `us-east-1`.

File: test/test_utils/__init__.py [hunk 7]
```diff
@@ -78,18 +78,38 @@ def get_ami_id_ssm(region_name, parameter_path):
     return ami_id
 
 
-# The Ubuntu 20.04 AMI which adds GDRCopy is used only for GDRCopy feature that is supported on PT1.13 and PT2.0
-UBUNTU_20_BASE_DLAMI_US_WEST_2 = get_ami_id_boto3(
-    region_name="us-west-2", ami_name_pattern="Deep Learning Base GPU AMI (Ubuntu 20.04) ????????"
+# DLAMI Base is split between OSS Nvidia Driver and Propietary Nvidia Driver. see https://docs.aws.amazon.com/dlami/latest/devguide/important-changes.html
+UBUNTU_20_BASE_OSS_DLAMI_US_WEST_2 = get_ami_id_boto3(
```
Comment (arjkesh) [inline test/test_utils/__init__.py:82]: looked at scope of removing these, and it will over-scope this PR. We can proceed with this for now

**Review by arjkesh — COMMENTED** [review]
Nice work and thorough testing

**Review by arjkesh — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Review by ztlevi — APPROVED** [review]
